### DISCUSSION TITLE
Error when connect to OPC UA server with long URL?

### DISCUSSION THREAD

**Discussion by nqduy35** [eclipse-milo/milo#1188] [General]
I'm using Eclipse Milo v0.6.11. I have two OPC UA server, one with a short URL: **opc.tcp://127.0.0.1:4840** and the other with a longer URL: **opc.tcp://127.0.0.1:4880/abc/xyz**.

I try with UaExpert and it works in both cases. However, when I try with Eclipse Milo, only the first case works. Here is the error when running the second case:
![2023-12-07-183633_1005x53_scrot](https://github.com/eclipse/milo/assets/3102822/149fbe72-cfb4-4656-8bd2-f75c96663227)

Do you have any idea to resolve this problem? Thank you in advanced.

**Comment by kevinherron** [comment]
That's not nearly enough information to help.

Please provide the full stack trace at a minimum, and a Wireshark capture to go with it would be ideal.

**Comment by nqduy35** [comment]
Thank @kevinherron for considering my question. Here is the Wireshark captures: 
[error-milo-1188.zip](https://github.com/eclipse/milo/files/13603709/error-milo-1188.zip)

You can find in this folder: (1) uaexpert good with short URL port 4840, (2) uaexpert good with long URL port 49300, (3) milo good with short URL port 4840, (4) milo BAD with long URL port 49300. 

- My silly thought: I'm not sure if milo read the part after the port. For example with URL **opc.tcp://127.0.0.1:4880/abc/xyz**, I'm wondering if it skips the part **/abc/xyz** or not?

**Reply by kevinherron** [reply]
I assume you mean port 59300, as that's what the captures show.

In the Milo capture the GetEndpoints response comes back with endpoint URLs like this: `opc.tcp://11f19fb339d6:59300/Prosyst/UnifiedDataCollector`.

The issue is probably that you can't resolve the hostname in those URLs. When the server returns endpoints with a different hostname than the one you used to connect you must modify the endpoints yourself before attempting to use them, especially if they contain a hostname you cannot resolve. This is what UaExpert is likely doing as well.

See https://stackoverflow.com/a/40555243

**Reply by nqduy35** [reply]
I've just tested the application today and everything works fine now. Thanks a lot.

**Comment by nqduy35** [comment]
Your analysis seems right. I will go for this direction. Thank you again for your support.
